File: src/test/java/amailp/intellij/robot/lexer/BaseLexerTest.java
package amailp.intellij.robot.lexer;

import com.intellij.lexer.Lexer;
import com.intellij.psi.tree.IElementType;
import org.junit.After;
import org.junit.Before;

import static org.hamcrest.CoreMatchers.*;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertThat;

public class BaseLexerTest {

    protected Lexer robotLexer;

    @Before
    public void initLexer() {
        robotLexer = new RobotLexer();
    }

    @After
    public void checkAllWasLexed() {
        assertEquals("Something was left unlexed", robotLexer.getTokenStart(), robotLexer.getTokenEnd());
    }

    protected void scanString(String toBeScanned) {
        robotLexer.start(toBeScanned);
    }

    protected void nextTokenIsType(IElementType type) {
        assertThat(type, equalTo(robotLexer.getTokenType()));
        robotLexer.advance();
    }

    protected void nextTokenIs(String text, IElementType type) {
        assertThat(text, equalTo(robotLexer.getTokenText()));
        nextTokenIsType(type);
    }
}

File: src/test/scala/amailp/intellij/robot/lexer/BaseLexerTest.scala
package amailp.intellij.robot.lexer

import org.junit.runner.RunWith
import org.scalatest.BeforeAndAfter
import com.intellij.psi.tree.IElementType
import org.scalatest.funsuite.AnyFunSuite
import org.scalatestplus.junit.JUnitRunner

@RunWith(classOf[JUnitRunner])
class BaseLexerTest extends AnyFunSuite with BeforeAndAfter {
  var robotLexer: RobotLexer = null

  before {
    robotLexer = new RobotLexer
  }

  after {
    def checkAllWasLexed() = assert(robotLexer.getTokenStart == robotLexer.getTokenEnd, "Something was left unlexed")
    checkAllWasLexed()
  }

  def scanString(toBeScanned: String): Unit = {
    robotLexer.start(toBeScanned)
  }

  def nextTokenIsType(elementType: IElementType): Unit = {
    assert(elementType == robotLexer.getTokenType)
    robotLexer.advance()
  }

  def nextTokenIs(text: String, elementType: IElementType): Unit = {
    assert(text == robotLexer.getTokenText)
    nextTokenIsType(elementType)
  }
}
